text/0039-source-removal-buildpack.md

```diff
@@ -0,0 +1,61 @@
+# Source Removal Buildpack
+
+## Summary
+
+Adopt [ForestEckhardt's Source Removal buildpack](https://github.com/ForestEckhardt/source-removal)
+into the Paketo Project to allow for uniform source removal logic across all
+buildpacks that require such functionality.
+
+## Motivation
+
+There are several buildpacks in the Paketo Buildpacks project that take their
+given source code and transform that into a new artifact such as a binary or
+other forms of compiled/transpiled code. It is often desirable to remove any
+unnecessary source code from the final image as it causes image bloat and poses
+a potential security risk. In order to currently accomplish this, any buildpack
+that wants to remove source code must have logic to remove it from the
+`workspace` so that it does not appear in the final image, however this task is
+complicated because often times there are static files that are required for
+the app to run that should no be deleted from the `workspace`. This leads to
+multiple implementations of non-trivial file removal logic across different
+buildpacks.
+
+This could all be simplified by adding a buildpack that performs all source
+removal logic to the end of any buildpack order that would require things to be
+removed from the `workspace` which is exactly what ForestEckhardt's Source
+Removal buildpack does. This buildpacks allows this shared logic to be
+centralized into one place and then reused across the entire buildpack
+ecosystem and as needed by particular users.
+
+## Rationale and Alternatives
+
+- Continue to implement source removal logic in all buildpacks that would
+  benefit from such functionality.
+
+## Implementation
+
+Currently ForestEckhardt's Source Removal buildpack always passes detection and
+has three actions.
+
+1. If no configuration is set (i.e. `BP_INCLUDE_FILES` and `BP_EXCLUDE_FILES`
+   are not set), then all the content of the `workspace` will be deleted.
+2. If `BP_INCLUDE_FILES` is set, then all files that **do not** match the
+   pattern globs specified in the environment variables will be deleted.
+3. If `BP_EXCLUDE_FILES` is set, then all files that **do** match the pattern
+   globs specified in the environment variables will be deleted.
+
```

Comment:
A couple of items that might be interesting for inclusion:

1. Support both `BP_INCLUDE_FILES` and `BP_EXCLUDE_FILES`. I have seen this with some tools, and I think it typically works by applying them in order, so you only include what matches `BP_INCLUDE_FILES` and from that subset of files, you apply the excludes filter.

    It's probably not as common of a use case, but I could potentially see where you might have a folder like `static/` where you only want to include that but then perhaps you want to exclude some content from under that? Maybe exclude hidden files?

2. Regex support. I know, I know. Regex just adds problems, lol, but it is a powerful way to match and customize filtering.

Perhaps #2 is too specific a detail. In which case, we perhaps reword the descriptions here such that they don't prescribe a specific matching implementation like glob or regex, but leave that up to the implementation.

Reply:
1. I think that is not a complete unreasonable request but I would like to see it for a user before complicating the logic more than it already is. I personally feel that this is something that could be very easily added after the buildpack has been adopted and shouldn't be a blocker.
2. I am reluctant to allow for users to user regex primarily because they are not really the convention used by other file control systems. I think I would have to hear a compelling reason to use regex over relatively standard file glob syntax.

Reply:
I'm Ok with suggesting them later.